Lightprotocol/light-protocol
# fix: Handle conflicting changelogs in concurrent Merkle tree

Restore the usage of `ConcurrentMerkleTreeError::CannotUpdateLeaf` error. Throwing an error when any changelog entry that modifies the same leaf we try to update is a right thing to do.

In case of indexed Merkle tree, we can ignore this error. We are already patching the leaf using the indexed changelog.

Other changes:

* Fix `test_100_nullify_mt`:
  * Each Merkle tree update was pointing to old `changelog_index`. And the crank Merkle tree was not being updated after each operation.
* Implement `Clone` in `BoundedVec`, to make cloning of Merkle proofs possible.
* Add more tests:
  * Check whether elements returned in `ConcurrentMerkleTree::changelog_entries`
    are as expected.
  * Test proving that we need to `skip(1)` during Merkle proof update.
  * Test with an update on wrapped-around changelog.
